Avoid reported unsoundness for implied lifetime bounds

YpeKingma opened the pull request:
changelog: [`lifetimes_bound_nested_ref`]: Avoid some unsoundness for implied lifetime bounds

The lints here are unusual in that they are related to a compiler bug.
This zulip discussion was my starting point:
https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/cve-rs

In the discussion, this overview on implied bounds was mentioned:
    https://github.com/rust-lang/rustc-dev-guide/blob/478a77a902f64e5128e7164e4e8a3980cfe4b133/src/traits/implied-bounds.md

and these are reported compiler bugs:

https://github.com/rust-lang/rust/issues/25860
Implied bounds on nested references + variance = soundness hole
https://github.com/rust-lang/rust/issues/25860#issue-82044592

https://github.com/rust-lang/rust/issues/84591
HRTB on subtrait unsoundly provides HTRB on supertrait with weaker implied bounds

https://github.com/rust-lang/rust/issues/100051
Implied bounds from projections in impl header can be unsound


One lint here suggests to add a generic lifetime bound for each function argument and return value
that is a nested reference with lifetimes, https://github.com/rust-lang/rust/issues/25860 .

This generic lifetime bound is implied by this nested reference,
the nested reference lifetime should outlive the initial reference lifetime.

Without the added lifetime bound the current compiler (1.76.0) can generate unsound code,
as reported above.
However when the generic lifetime bound implied by the nested reference is added,
the compiler produces an error message when the function is used.
Therefore this lint can help to avoid the unsoundness in this case.

The other lint here is the reverse of the first one.
It suggests to delete this generic lifetime bound because it is implied.
This lint can help to clean up code, but only after the corresponding unsoundness is fixed
in the compiler.
I have added the reverse lint because of the unusual situation here.

For the other two cases
    https://github.com/rust-lang/rust/issues/84591 and
    https://github.com/rust-lang/rust/issues/100051 ,
the situation is similar: when manually adding a generic lifetime bound that is implied
by a nested reference, the compiler gives an error instead of producing unsound code.

Pending the review here, I want to continue with similar pairs of lints for these other two cases.

Is it necessary to document the error messages that the compiler gives when 
the generic lifetime bounds are added in the reported examples?

Manishearth commented:
r? @oli-obk would you be willing to review this?

oli-obk reviewed (changes requested):
I think this will be too noisy to use, because it's only a problem when the lifetimes are used again in other arguments and return types, and I don't actually know the details of where it will be problematic. Maybe at minimum check that the lifetimes are used again at all

YpeKingma commented:
> I think this will be too noisy to use, because it's only a problem when the lifetimes are used again in other arguments and return types, and I don't actually know the details of where it will be problematic.

Adding any number of implicit lifetime bounds is correct (but should not be necessary, see the reverse lint.)
When this happens too often it could become noisy, and then this lint might be extended to automatically apply the suggested addition.

I could try check how often this occurs by using cargo dogfood. I suppose that includes the source code of clippy,
and I see plenty of lifetimes in there. 

The unknown problem in this case is the possible unsoundness when an implicit lifetimes bound is added and the compiler does not complain about the new code.

> Maybe at minimum check that the lifetimes are used again at all

Checking for lifetimes that are not used again at all sounds like another lint, but I may misunderstand this.
I suppose this can occur for pub functions/... that are not used elsewhere.

In the reported pieces of code, after adding an implicit lifetime the compiler gives an error at later use.
A check for the possibility of a later compiler error could be added in the clippy run, but that might lead to false negatives when this check is not implemented correctly.

YpeKingma commented on clippy_lints/src/lifetimes_bound_nested_ref.rs:
```diff
@@ -0,0 +1,222 @@
+    let TyKind::Ref(mut lifetime, mut mut_ty) = ty.kind else {
+        return implied_bounds;
+    };
+    while let TyKind::Ref(nested_lifetime, nested_mut_ty) = mut_ty.ty.kind {
```
> also this doesn't handle type aliases. It may be easier to do this analysis on the `ty` datastructures instead of on the `hir` datastructures. You can get the `ty::FnSig` via `cx.tcx.fn_sig(local_def_id)`.

I'll try and use the ```ty::FnSig``` instead of the things from ```hir``` to collect the implied bounds. In the example ``` &'a (&'b i32, u32)``` I see that this will probably need recursion, and implementing that might take a while.

oli-obk commented:
> Checking for lifetimes that are not used again at all sounds like another lint, but I may misunderstand this.

I mean to not emit the lint unless the lifetime appears twice in the signature.

As far as I understood it, that is how you can get the unsoundness. Not with a function that's just missing an implied bound on two lifetimes, but additionally has one (or both?) Lifetimes again elsewhere in the same signature.

I don't have a strong opinion on whether to land the lint as is or improve it. I think you'd want to first run it on a few crates to see if it is useful in practice, or if it mainly asks to annotate items that don't have any soundness concerns.

It's fine to land with lots of missed situations. I just worry about lots of false alarms.

oli-obk commented on clippy_lints/src/lifetimes_bound_nested_ref.rs:
```diff
@@ -0,0 +1,222 @@
+                    ctx,
+                    ADD_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+                    generics.span,
+                    &format!(
```
The one thing I think should happen before landing this is to make it a structured suggestion, so people can just rustfix their crates instead of having to manually adjust possibly hundreds of functions

flip1995 commented on clippy_lints/src/lifetimes_bound_nested_ref.rs:
```diff
@@ -0,0 +1,222 @@
+/// Lints to help dealing with unsoundness due to a compiler bug described here:
```
Starting a thread to localize this discussion:

My question is about the motivation of including this into Clippy: If Clippy is able to detect those unsoundness issues, shouldn't this be implemented in the compiler and fix the unsoundness bugs?

One explanation for this approach that I see is, that this lint will have false positives (FPs), so implementing it in the compiler would restrict the language too much. Am I right in that assumption?

If this should be the case, putting the lint into a warn-by-default group might be hard. We definitely should run our `lintcheck` tool to look for FPs in popular crates.

YpeKingma replied:
> Starting a thread to localize this discussion:
> 
> My question is about the motivation of including this into Clippy: If Clippy is able to detect those unsoundness issues, shouldn't this be implemented in the compiler and fix the unsoundness bugs?

Ideally yes, and I hope the compiler gets fixed before this is ready. 
But the 25860 issue was reported in 2015...

> 
> One explanation for this approach that I see is, that this lint will have false positives (FPs), so implementing it in the compiler would restrict the language too much. Am I right in that assumption?

An implied bound is correct because it is implied by the compiler input code. Adding an implied bound is also correct, but it adds redundancy in the input code. I think there is no language restriction here.

oli-obk replied:
Yea, this isn't implemented in rustc, because it would require bounds in a lot of places, while the new solver that can fix this (citation needed) is still about a year out.

I just wonder how many ppl will use this lint. I don't want you to waste time writing a complex lint that then becomes mostly unused

flip1995 commented on clippy_lints/src/lifetimes_bound_nested_ref.rs:
```diff
@@ -0,0 +1,222 @@
+    /// ### Known problems
+    /// It is not known whether this covers all cases in issue 25860.
+    ///
+    /// ### Example, the val_a argument implies a lifetimes bound:
```
NIT:

```suggestion
    /// ### Example
    /// The `val_a` argument implies a lifetimes bound:
```

flip1995 commented on clippy_lints/src/lifetimes_bound_nested_ref.rs:
```diff
@@ -0,0 +1,222 @@
+    /// should only be done after the compiler
+    /// has been fixed to deal correctly with implied lifetime bounds.
+    ///
+    /// ### Example, the val_a argument implies a lifetimes bound:
```
```suggestion
    /// ### Example
    /// The `val_a` argument implies a lifetimes bound:
```

flip1995 commented on clippy_lints/src/lifetimes_bound_nested_ref.rs:
````diff
@@ -0,0 +1,222 @@
+    /// pub const fn lifetime_translator<'a, 'b: 'a, T>(val_a: &'a &'b (), val_b: &'b T) -> &'a T {...}
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub ADD_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
````
Those lint names are a mouthful.

Maybe: `IMPLICIT_LIFETIME_BOUNDS` here and `EXPLICIT_*` for the lint that removes those again?

YpeKingma replied:
> Those lint names are a mouthful.
> 
> Maybe: `IMPLICIT_LIFETIME_BOUNDS` here and `EXPLICIT_*` for the lint that removes those again?

From back to front:

The ```_FN``` suffix is there to indicate to the 25860 issue. The idea was to start of with different suffixes for the two issues but merging the lints into the same name makes them easier to use, so I'll drop the suffix.

```LIFETIME_BOUNDS``` is easier to read by not accurate: the lint gives an error message for a single bound, 
and each bound has two lifetimes.

I'll change the lint names to use ```IMPLICIT``` instead of ```ADD_REDUNDANT```, and similarly for the explicit and remove redundant.

flip1995 replied:
> `LIFETIME_BOUNDS` is easier to read by not accurate

Please refer to the [lint naming conventions](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints), 2. paragraph:

> The basic rule is: the lint name should make sense when read as "allow lint-name" or "allow lint-name items". For example, "allow deprecated items" and "allow dead_code" makes sense, while "allow unsafe_block" is ungrammatical (should be plural).

---

I think you can keep the `_FN` suffix to separate those lints, but I'm not convinced that the `NESTED_REF` really has to be part of the lint name.

bors commented:
:umbrella: The latest upstream changes (presumably #12451) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

YpeKingma commented:
Well, after some back and forth with CI/CD I don´t know the root cause of the last error in the Clippy Test / base (pull_request) above. It's in the code docs, so I'll leave it at that for now.

The actual code is for issue 25860:

I renamed the lints to IMPLICIT_LIFETIMES_BOUND_NESTED_REF and  EXPLICIT_LIFETIMES_BOUND_NESTED_REF.
Still a mouth full, but a bit shorter than earlier with ADD_IMPLIED.... and REMOVE_EXPLICIT... .
Maybe with the other two cases (see below) added, the suffix _REF can be dropped.

The implicit lifetime bounds are now derived from the the fn_sig() as suggested. 
The recursion on Ty for this worked out well, but perhaps it can be done by a visitor instead,
see ```collect_nested_ref_implied_bounds()```.

For the rest, I changed the lifetime bound pair struct to use String for simplicity.
This assumes that the lifetime names are unique, i.e. they should be taken from a single scope.

Not yet done:

Check a reported lifetime bound has a lifetime that occurs at least twice in the function signature.  
Should that depend on which lifetime (long/outlived) is involved?

Make the lints a structured suggested to allow automatic fixing, both ways. 
This will need better dealing with the source spans, currently on the generics span is used.
The source span for an implied nested reference could also be useful in the error message.

Remaining cases for lints:
```
// from https://github.com/rust-lang/rust/issues/100051
impl<'a: 'b, 'b> Extend<'a, 'b> for <&'b &'a () as Trait>::Type {...}
// impl<'a, 'b> Extend<'a, 'b> for <&'b &'a () as Trait>::Type {...}


// from https://github.com/rust-lang/rust/issues/84591
impl<'a: 'b, 'b> Subtrait<'a, 'b, &'b &'a ()> for MyStruct {}
// impl<'a, 'b> Subtrait<'a, 'b, &'b &'a ()> for MyStruct {}
```

How much code documentation is needed?

YpeKingma commented:
The commits today for the other two went faster than I expected. Please include them for review.

YpeKingma commented:
cargo dev dogfood does not pass here, fixing...

YpeKingma commented:
cargo uitest fails nicely on useless_asref.rs line 23:
```rust
impl<'a, 'b, 'c> AsRef<&'a &'b &'c MoreRef> for MoreRef
```
with two error messages for missing implied bounds  ```'b: 'a ``` and ```'c, 'b```.
I don't think that needs to be corrected.

I also made a local version with the IMPLICIT* lint in the category suspicious instead of nursery.
```cargo lintcheck``` with local version did not make the ```clippy::implicit-lifetimes-bound-nested-ref``` lint show up in the stats,
so it could be that there is no noise from there.

YpeKingma commented:
A possible way to assign these lints to categories over a longer periods of time, several releases around the compiler fix for the related unsoundness:
- implicit-lifetimes-bound-nested-ref : first nursery to try, then pedantic, then (may be) suspicious, then compiler fix, then deprecated.
- explicit-lifetimes-bound-nested-ref : first nursery to wait, then compiler fix, then complexity.

The pedantic category fits nicely with the implication here.

YpeKingma commented:
The Clippy Test / base is failing because I missed a cargo uibless. Fixing...

YpeKingma commented:
Rustc was updated to 1.77.0 today, and I'm using it now.
How do I deal with this update here? 
Can I merge master into the lifetimes_bound_nested_ref branch, or should I squash the update here?

flip1995 commented:
Running this on the top 500 crates by downloads, triggered the lint 4 times:

clippy 0.1.78 (87325570c61 2024-03-20)

### Reports

target/lintcheck/sources/clap_builder-4.5.2/src/parser/features/suggestions.rs:38:32 clippy::implicit_lifetimes_bound_nested_ref "missing lifetime bound declation: '_: '_"
target/lintcheck/sources/httparse-1.8.0/src/lib.rs:903:22 clippy::implicit_lifetimes_bound_nested_ref "missing lifetime bound declation: '_: '_"
target/lintcheck/sources/httparse-1.8.0/src/lib.rs:916:27 clippy::implicit_lifetimes_bound_nested_ref "missing lifetime bound declation: 'b: 'a"
target/lintcheck/sources/httparse-1.8.0/src/lib.rs:936:29 clippy::implicit_lifetimes_bound_nested_ref "missing lifetime bound declation: '_: '_"


### Stats:

| lint                                               | count |
| -------------------------------------------------- | ----- |
| clippy::implicit_lifetimes_bound_nested_ref        |     4 |


---

I don't have the time to look into those rn.

YpeKingma commented:
> Running this on the top 500 crates by downloads, triggered the lint 4 times:
> 
...

Thank you.

The reported bounds with empty lifetime names indicate a bug here. 
I'll get these two crates to use them as test cases, and then:
- try and recompile after manually adding this reported bound: ```httparse-1.8.0/src/lib.rs:916:27```
- investigate the reporting of empty lifetime names.

YpeKingma commented:
This is most likely the source code for the reported missing bounds (currently at line 1027 in src/lib.rs of the httparse crate):
```
unsafe fn deinit_slice_mut<'a, 'b, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
    let s: *mut &mut [T] = s;
    let s = s as *mut &mut [MaybeUninit<T>];
    &mut *s
}
```

Since the argument and return type have the same reference structure, I would not expect this to cause unsoundness.

Shall I add a condition for the lint here that the implied bounds from input arguments and from the return type are not the same?

flip1995 commented:
> Shall I add a condition for the lint here that the implied bounds from input arguments and from the return type are not the same?

If it can't cause unsoundness issues and your motivated to add a check for it, I'd say sure, go for it.

---

Sidenote: there's a typo in the error message: `declation`

YpeKingma commented:
> This is most likely the source code for the reported missing bounds (currently at line 1027 in src/lib.rs of the httparse crate):
> 
> ```
> unsafe fn deinit_slice_mut<'a, 'b, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
>     let s: *mut &mut [T] = s;
>     let s = s as *mut &mut [MaybeUninit<T>];
>     ype@v110:~/gitrepos/httparse$ git diff -U0

I added the implied bound manually on its current master, and it still compiles and tests cleanly:

```
diff --git a/src/lib.rs b/src/lib.rs
index 3232852..457f225 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1027 +1027 @@ fn parse_headers_iter<'a>(
-unsafe fn deinit_slice_mut<'a, 'b, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
+unsafe fn deinit_slice_mut<'a, 'b: 'a, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
&mut *s
> }
> ```

YpeKingma commented:
> > Shall I add a condition for the lint here that the implied bounds from input arguments and from the return type are not the same?
> 
> If it can't cause unsoundness issues and your motivated to add a check for it, I'd say sure, go for it.

I'm not sure, I have no idea how lifetimes are currently implemented.

> 
> Sidenote: there's a typo in the error message: `declation`

Fixed here, will be in the next push.

YpeKingma commented:
Today's push contains:

Correct the declation typo in the error message.

Add a test case with equal implied bounds from function inputs and return.
Do not test for this equality, i.e. the lints here are activated for this case.
This testcase is there to get more opinions on it.

Use lifetimes found on ADT's, i.e. on struct/union/enum.
This adds quite a few lint activations here.

Only use declared lifetimes for an implicit bound.
This fixes the bug using the '_ lifetime. 

Only use different lifetimes for an implicit bound, this was a bug.

Use a local stack to avoid most recursion over types.
Introduce a linting struct to avoid passing many arguments.
Use FxHashSet as suggested by a compiler lint.

Was merged with master of 24 March 2024.

Local test results, with implicit-lifetimes-bound-nested-ref in the suspicious category (only locally here, the lints are in nursery):

cargo dev dogfood gives 13 lints, most of them suggesting to add the bound 'tcx: 'a, or 'hir: 'a .

cargo lintcheck gives 2 lints:
target/lintcheck/sources/serde-1.0.118/src/private/de.rs:1771:29 clippy::implicit_lifetimes_bound_nested_ref "missing lifetime bound declaration:
 'de: 'a"
target/lintcheck/sources/serde-1.0.118/src/ser/fmt.rs:37:5 clippy::implicit_lifetimes_bound_nested_ref "missing lifetime bound declaration: 'b: '
a"

YpeKingma commented:
On the allow convention for lint names. Currently: 

the implicit-lifetimes-bound-nested-ref lint suggests to add an implicit bound, and

the explicit-lifetimes-bound-nested-ref lint suggests to remove it.


That is actually contrary to this convention for lint names. Shall I exchange and simplify the lint names, so they become:

explicit-lifetimes-bound to allow an implicit bound as explicit, and suggest to add the implicit bound as explicit, and

implicit-lifetimes-bound to allow implicit bounds only, and suggest to remove and explicit bound when it is implicit?

YpeKingma commented:
Latest push:

Rename lints to be more in line with the "allow" naming convention.
This involved exchanging the two names, and I also made the names shorter.
I revised the documentation for the lints and tried to make it consistent with this rename.
This is the code with the lint names:
```rust
impl_lint_pass!(LifetimesBoundNestedRef => [
    EXPLICIT_LIFETIMES_BOUND,
    IMPLICIT_LIFETIMES_BOUND,
]);
```
The explicit one allows and suggests explicit declarations of lifetime bounds.
The implicit one allows implicit bounds only, and suggests to remove the declarations.

The explicit lint to add explicit declarations of lifetime bounds is now machine applicable.
The other lint to remove these declarations is not machine applicable, and I think it will be good to keep it that way.
I tried testing both together with a .fixed file, but that won't work with two lints in opposite directions.
The situation is unusual, and in case it is preferred to have the fixing tested I could use
some help on how to get this done. Perhaps use different .rs modules for the lints, so the tests can be split? 

The type walking to collect implicit bounds now also does the dyn Dynamic types.

y21 commented:
> I tried testing both together with a .fixed file, but that won't work with two lints in opposite directions.
The situation is unusual, and in case it is preferred to have the fixing tested I could use
some help on how to get this done. Perhaps use different .rs modules for the lints, so the tests can be split?

This is the usual, preferred way, yes. One uitest should always test at most one lint

YpeKingma commented:
The push of 2 April adds a lint note to the nested reference that implies the lifetimes bound.
To have this I had to use an early lint pass instead of the previous late lint pass, and that affects a lot of the implementing code.
The test cases input source code has not changed, only the expected stderr now also contains a lint note for the nested reference.

The results of cargo dev dogfood test and of cargo lintcheck of 25 March did not change either.
These use the implied_bound_explicit lint in the suspicious category.
The pull request has both lints in the nursery category.

YpeKingma commented:
There are some CI errors in unrelated code.

In today's push I mostly improved the docs. There were some errors in doc url's and I hope these are corrected now.
Also I tried to make the help text for the lints be clear about how to use the lints: 
- after declaring an implicit lifetimes bound, other code many need correction,
- do not remove declared implicit lifetime bounds until the compiler is fixed to use the implied bounds correctly.

Is there a way for me to check the generated help text of the lints?

xFrednet commented:
Hey @YpeKingma , this is a ping from triage, since there hasn't been any activity in some time. Are you still planning to continue this implementation?

If you have any questions, you're always welcome to ask them in this PR or on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy).

@rustbot author

YpeKingma commented:
I'd like to know what still needs to be done to have this accepted. I don't know how to deal with the remaining CI errors, so any tips/hints are welcome.

For now, I can only follow rustc releases, and I'm planning to do an update here for rustc 1.79.0 in about three weeks.

Also this may become superfluous because of rustc improvements in dealing with the unsafety that this tries to avoid. I have no idea how quickly rustc is going there.

flip1995 commented:
I think I need to review this PR in depth. I try to get to it this weekend, otherwise I'll reassign.

r? flip1995 

@rustbot ready

xFrednet commented:
Ping @flip1995, do you still want to give this PR a review or reroll?

flip1995 commented:
Let's reroll. I enjoyed the nice weather over the weekend instead and won't have time this week either.

r? clippy

I really want to give this another look though. But merging shouldn't be blocked by me.

Manishearth commented:
r? clippy

also rerolling, tad busy today

YpeKingma commented:
On today's push: see the 3 commit messages.

Only one CI failure is left, and it indicates a missing an unclosed delimiter for the docs of main function.  However I still cannot find the root cause for this failure.

YpeKingma commented:
Today's push: no functional changes, only updates to follow the master branch.

Meanwhile there were some comments on the unsoundness bugs mentioned above.

I think the best next step would to build this into the compiler to fix these bugs.
But I have no idea how to start with that, and I don't know how these recent comments relate to that.

YpeKingma commented:
r? @llogiq

YpeKingma commented:
I'll try and update this to upstream master tomorrow at the Rustweek hackathon day.

YpeKingma commented:
I will archive my repo https://github.com/YpeKingma/rust-clippy in week or so, and close this PR just before that.

In case you want to take this further please use the branch with the latest development code:

https://github.com/YpeKingma/rust-clippy/tree/lifetimes_bound_nested_ref_dev

and skip the last commit there, just use: 

https://github.com/rust-lang/rust-clippy/commit/da97970629b4f2edf3cbf554aa39aa498f057334


On that commit there is a (probably minor) problem: after ```cargo uibless```, ```cargo uitest``` does not pass, which contradicts the clippy development docs.
